opcua/107-opcuamethod.js

```diff
@@ -183,39 +184,114 @@ module.exports = function (RED) {
+      if (node.client) {
+        node.client = null;
+      }
       try {
-        // step 1 : connect to
-        await client.connect(url);
-        node.log("start method client on " + opcuaEndpoint.endpoint);
-
-        // step 2 : createSession
-        const session = await client.createSession(userIdentity);
-        verbose_log("start session on " + opcuaEndpoint.endpoint);
-        set_node_status_to("session active");
-        node.session = session;
-        verbose_log("Call method: " + JSON.stringify(message));
-        var status = await callMethod(message);
-        
-        if (node.session) {
-          await node.session.close();
-          verbose_log("Session closed");
-          node.session = null;
-          await client.disconnect();
+        if (opcuaEndpoint.endpoint.indexOf("opc.tcp://0.0.0.0") == 0) {
+          set_node_status_to("no client");
+          if (callback) {
+            callback();
+          }
+          return;
         }
-        if (status === opcua.StatusCodes.Good) {
+        // Normal client
+        verbose_log(chalk.green("1) CREATE CLIENT: ") + chalk.cyan(JSON.stringify(connectionOption).substring(0,75) + "..."));
+        node.client = opcua.OPCUAClient.create(connectionOption);
+        node.client.on("connection_reestablished", reestablish);
+        node.client.on("backoff", backoff);
+        node.client.on("start_reconnection", reconnection);
+      }
+      catch(err) {
+        node_error("Cannot create client, check connection options: " + stringify(connectionOption));
+      }
+      set_node_status_to("create client");
+      if (callback) {
+        callback("");
+      }
+    }
+
+    create_opcua_client(function (err) {
+        if (err) {
+          node_error(err);
+          node.status({
+            fill: "red",
+            shape: "dot",
+            text: "Error: " + err.toString()
+          });
+        }
+      });
+
+    set_node_status_to("initialized");
+
+    async function methodNodeProcess(url, message, callback) {
+      try {
+
+        const statuses = ['initialized', 'method executed'];
+
+        verbose_log("Queued Message: " + JSON.stringify(message));
+        cmdQueue.push(message);
+
+        if (statuses.includes(currentStatus)) {
           node.status({
             fill: "green",
             shape: "dot",
-            text: "Method executed"
+            text: "Executing method"
           });
+          // if Node Re-connecting or busy, methods should only be queued
+          // If it's ready:
+
+          // step 1 : connect client
+          await connect_opcua_client(url);
+          node.log("start method client on " + opcuaEndpoint.endpoint);
+
+          // step 2 : createSession
+          node.session = await node.client.createSession(userIdentity);
+          verbose_log("start session on " + opcuaEndpoint.endpoint);
+          set_node_status_to("session active");
+
+          // step 3: call method
+          for (var i in cmdQueue) {
```

Comment:
```suggestion
          for (const cmd of cmdQueue) {
```
i think it's much more elegant, and you should never use var if you can help it anyway. at least you could change it to
```suggestion
          for (const i in cmdQueue) {
```

Reply:
Done